v2/tools/generator/internal/conversions/property_conversions.go

```diff
@@ -1374,73 +1391,162 @@ func assignObjectFromObject(
-			actualReader = star.X
-		} else {
-			actualReader = astbuilder.AddrOf(reader)
-		}
-
-		var functionName string
-		var conversion dst.Stmt
-		if conversionContext.direction == ConvertFrom {
-			// Destination is our current type
-			functionName = NameOfPropertyAssignmentFunction(sourceName, ConvertFrom, conversionContext.idFactory)
-			conversion = astbuilder.AssignmentStatement(
-				errLocal,
-				tok,
-				astbuilder.CallExpr(localId, functionName, actualReader))
-		} else {
-			// Destination is another type
-			functionName = NameOfPropertyAssignmentFunction(destinationName, ConvertTo, conversionContext.idFactory)
-			conversion = astbuilder.AssignmentStatement(
-				errLocal,
-				tok,
-				astbuilder.CallExpr(reader, functionName, astbuilder.AddrOf(localId)))
-		}
-
-		transition := sourceEndpoint.Name()
-		if sourceEndpoint.Name() != destinationEndpoint.Name() {
-			transition = fmt.Sprintf("%s from %s", destinationEndpoint.Name(), sourceEndpoint.Name())
-		}
+		functionName := NameOfPropertyAssignmentFunction(destinationName, ConvertTo, conversionContext.idFactory)
+		conversion := astbuilder.AssignmentStatement(
+			errLocal,
+			tok,
+			astbuilder.CallExpr(reader, functionName, astbuilder.AddrOf(localId)))
 
 		checkForError := astbuilder.ReturnIfNotNil(
 			errLocal,
 			astbuilder.WrappedErrorf(
 				errorsPackageName,
-				"calling %s() to populate field %s", functionName, transition))
+				"calling %s() to %s",
+				functionName,
+				describeAssignment(sourceEndpoint, destinationEndpoint)))
 
 		assignment := writer(dst.NewIdent(copyVar))
 		return astbuilder.Statements(declaration, conversion, checkForError, assignment)
 	}, nil
 }
 
+// assignObjectsViaIntermediateObject will generate a conversion if both properties are TypeNames referencing ObjectType
+// definitions, neither property is optional, and the types are NOT adjacent in our storage conversion graph.
+// The conversion is implemented by assigning properties to an intermediate instance before assigning those to our
+// actual destination instance.
+//
+// For ConvertFrom the generated code will be:
+//
+// var <local> <intermediateType>
+// err := <local>.AssignPropertiesFrom(<source>)
+// if err != nil {
+//     return errors.Wrap(err, "while calling <local>.AssignPropertiesFrom(<source>)")
+// }
+//
+// var <otherlocal> <destinationType>
+// err := <otherlocal>.AssignPropertiesFrom(<local>)
+// if err != nil {
+//     return errors.Wrap(err, "while calling <otherlocal>.AssignPropertiesFrom(<local>)")
+// }
+//
+// Note the actual steps are generated by nested conversions; this handler works by finding the two conversions needed
+// given our intermediate type and chaining them together.
+//
+func assignObjectsViaIntermediateObject(
+	sourceEndpoint *TypedConversionEndpoint,
+	destinationEndpoint *TypedConversionEndpoint,
+	conversionContext *PropertyConversionContext) (PropertyConversion, error) {
+
+	// Require both source and destination to not be bag items
```

Comment:
Consider putting these checks in a function to be re-used by other methods(assignObjectDirectlyFromObject) as well?

Reply:
I don't know that we can make these much smaller - we'd still need to call the function and then return `nil, nil` if it returned true, and possibly at the risk of obscuring which prerequisites are checked.  Moving the older checks into the `IsBagItem()` and `IsOptional()` predicates seemed like a good middle path.